src/replication.c

```diff
@@ -2309,258 +2233,464 @@ void readSyncBulkPayload(connection *conn) {
+         * order to detect timeouts during replication). */
+        atomic_store_explicit(&server.repl_transfer_lastio, atomic_load_explicit(&server.unixtime, memory_order_relaxed), memory_order_relaxed);
+
+        /* Write what we got from the socket to the dump file on disk */
+        if ((nwritten = write(server.repl_transfer_fd, buf, nread)) != nread) {
+            replicaBioSaveServerLog(LL_WARNING,
+                                    "Write error or short write writing to the DB dump file "
+                                    "needed for PRIMARY <-> REPLICA synchronization: %s",
+                                    (nwritten == -1) ? strerror(errno) : "short write");
+            error = 1;
+            goto done;
+        }
+        server.bio_repl_transfer_read += nread;
+
+        /* Delete the last 40 bytes from the file if we reached EOF. */
+        if (usemark && eof_reached) {
+            if (ftruncate(server.repl_transfer_fd, server.bio_repl_transfer_read - RDB_EOF_MARK_SIZE) == -1) {
+                replicaBioSaveServerLog(LL_WARNING,
+                                        "Error truncating the RDB file received from the primary "
+                                        "for SYNC: %s",
+                                        strerror(errno));
+                error = 1;
+                goto done;
+            }
+        }
+
+        /* Sync data on disk from time to time, otherwise at the end of the
+         * transfer we may suffer a big delay as the memory buffers are copied
+         * into the actual disk. */
+        if (server.bio_repl_transfer_read >= repl_transfer_last_fsync_off + REPL_MAX_WRITTEN_BEFORE_FSYNC) {
+            off_t sync_size = server.bio_repl_transfer_read - repl_transfer_last_fsync_off;
+            rdb_fsync_range(server.repl_transfer_fd, repl_transfer_last_fsync_off, sync_size);
+            repl_transfer_last_fsync_off += sync_size;
+        }
+
+        /* Check if the transfer is now complete */
+        if (!usemark) {
+            if (server.bio_repl_transfer_read == server.bio_repl_transfer_size) eof_reached = 1;
+        }
+    } while (!eof_reached);
+
+done:
+    /* Restore the socket to the original state to continue
+     * with normal replication. */
+    connNonBlock(conn);
+    connRecvTimeout(conn, 0);
+
+    /* We will not use the primary's connection any more, it is now safe to restore the main thread's access to it */
+    if (is_dual_channel) {
+        server.repl_rdb_transfer_s = conn;
+    } else {
+        server.repl_transfer_s = conn;
+    }
+
+    /* Handle completion */
+    if (error) {
+        replicaBioSaveServerLog(LL_WARNING, "Error downloading RDB");
+        atomic_store_explicit(&server.replica_bio_disk_save_state, REPL_BIO_DISK_SAVE_STATE_FAIL, memory_order_release);
+    } else {
+        replicaBioSaveServerLog(LL_NOTICE, "Done downloading RDB");
+        server.replica_bio_disk_save_conn = conn;
```

Comment:
can you explain why do we need this? we already restored the repl_transfer_s/repl_rdb_transfer_s so what is the point of remembering another connection?

Reply:
The RDB load part of the sync is handled by the main thread in `handleBioThreadFinishedRDBDownload` which is called asynchronously in the replication cron. That function needs to use that connection but it does not "know" whether we had a single or dual channel sync.

In other words - in the beginning of the sync we took away the connection from the main thread and gave it to the Bio thread. This is our way to give it back to the main thread.

Reply:
I understand the second part (even though I do not understand why we keep it this way). I still do not understand why we need to return it.

is it to perform this logic?:

```
/* Final setup of the connected replica <- primary link */
    if (conn == server.repl_rdb_transfer_s) {
        dualChannelSyncHandleRdbLoadCompletion();
    } else {
        replicationCreatePrimaryClient(server.repl_transfer_s, rsi->repl_stream_db);
        server.repl_state = REPL_STATE_CONNECTED;
        server.repl_down_since = 0;
        /* Send the initial ACK immediately to put this replica in online state. */
        replicationSendAck();
    }
```